appservice/multiplexer.py

```diff
@@ -285,6 +298,18 @@ async def watch_redis(channel):
                         logger.warning('invalid JSON, starting without sessions: %s', e)
                         SESSIONS = {}
 
+                    # resolve wait-running futures
+                    logger.debug('watch_redis WAIT_RUNNING_FUTURES before: %s', WAIT_RUNNING_FUTURES)
+                    for sessionid, wait_futures in WAIT_RUNNING_FUTURES.items():
+                        if SESSIONS.get(sessionid, {}).get('status') != 'running':
+                            continue
+
+                        for f in wait_futures.copy():
```

Comment:
Can you explain why the .copy() is required here?

Reply:
Thinking about it more, it's for race conditions?

Reply:
Because it's not safe to iterate over a list while you remove things from it, in line 320. The standard pattern is to copy the list first.